doc/macro_overview.md

```diff
@@ -0,0 +1,385 @@
+
+# Macro Overview
+
+The offical paper describing the mechanics behind Lean 4's macro system can be
+found in [Beyond Notations: Hygienic Macro Expansion for Theorem Proving
+Languages](https://arxiv.org/abs/2001.10490) by Sebastian Ullrich and Leonardo
+de Moura, and the accompanying repo with example code can be found in the
+paper's code [supplement](https://github.com/Kha/macro-supplement). The
+supplement also includes a working implementation of the macro expander, so it's
+a good case study for people interested in the details.
+
+## What is a macro in Lean?
+
+A macro is a function that takes in a syntax tree and produces a new syntax
+tree. Macros are useful for many reasons, but two of the big ones are a)
+allowing users to extend the language with new syntactic constructs without
+having to actually expand the core language, and b) allowing users to automate
+tasks that would otherwise be extremely repetitive, time-consuming, and/or
+error-prone.
+
+A motivating example is set builder notation. We would like to be able to write
+the set of natural numbers 0, 1, and 2 as just `{0, 1, 2}`. However, Lean does
+not natively support this syntax, and the actual definition of a set in Mathlib
+does not let us just declare sets in this manner; naively using the set API
+would force us to write `Set.insert 1 (Set.insert 2 (Set.singleton 3))`.
```

Comment:
Sorry for being very late to the conversation, but this wording seems awkward now that this documentation is part of the Lean core rather than Mathlib.

Reply:
I did think about replacing this with the `List` literal macro in the prelude, but it's not as simple as the `Set` one. If you can find a better example from the prelude that demonstrates a similar idea you're welcome to change it; I think this isn't terrible since the macro is recreated in full below the text.

Reply:
Sorry, my point was more the mention of 'the actual definition of a set in Mathlib` -- it is somewhat awkward mention a third party library in the Lean core's documentation. Admittedly, the `Set` part is also a probably a little confusing to CS people as they would probably think of the `RBSet` /  `HashSet` from `Std` in the Lean core first, which is a very different kind of set than the `Set` from mathlib. 

I am little curious, though, why do you think the Set macro is simpler than the List one? They both can expand similarly `{0,1,2}` to `(Set.inserrt 1 (Set.insert 2 (Set.insert 3 ∅)` and `[0,1,2]` to `List.cons 1 (List.cons 2 (List.cons 3 ∅)`.

Reply:
They expand similarly, but the definition of the list one is more complex than the set one: https://github.com/leanprover/lean4/blob/653ff184a817572b13d9a2a97aede3a3de7e925e/src/Init/Notation.lean#L190

Reply:
True, the actual implementation is  more complex (for optimization reasons), but for demonstration purposes they could be written essentially the same way.

Reply:
I agree with the overall point you're making; if you want to swap the `Set` thing out to talk about/demo a simplified version of the list macro I'm all for it.